server/conn.go

```diff
@@ -715,6 +775,32 @@ func (cc *clientConn) openSessionAndDoAuth(authData []byte) error {
 	return nil
 }
 
+func (cc *clientConn) checkAuthPlugin(ctx context.Context, authPlugin *string) ([]byte, error) {
```

Comment:
So, there are:
1. the default auth plugin (we set in the initial handshake packet)
2. the user's auth plugin, which stores in mysql.user authentication related column
3. the auth plugin from the client response packet

Which one is which? I'm a bit confused.

Pass *string as both input and output of a function is an idiom in C.
Go support multiple return, you don't need to do that.